hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java

```diff
@@ -3409,19 +3464,55 @@ public ShortCircuitRegistry getShortCircuitRegistry() {
   public void checkDiskError() throws IOException {
     Set<FsVolumeSpi> unhealthyVolumes;
     try {
-      unhealthyVolumes = volumeChecker.checkAllVolumes(data);
+      // check all volume
+      unhealthyVolumes = volumeChecker.checkAllVolumes(allData);
       lastDiskErrorCheck = Time.monotonicNow();
     } catch (InterruptedException e) {
-      LOG.error("Interruped while running disk check", e);
+      LOG.error("Interrupted while running disk check", e);
       throw new IOException("Interrupted while running disk check", e);
     }
 
     if (unhealthyVolumes.size() > 0) {
+      if (errorDisk == null) {
+        errorDisk = new ArrayList<>();
+      }
+      List<StorageLocation> tmpDisk = Lists.newArrayList(errorDisk);
+      errorDisk = null;
+      for (FsVolumeSpi vol : unhealthyVolumes) {
+        LOG.info("Add error disk " + vol.getStorageLocation()
+            + " to errorDisk - " + vol.getStorageLocation());
+        errorDisk.add(vol.getStorageLocation());
+        if (tmpDisk.contains(vol.getStorageLocation())) {
+          tmpDisk.remove(vol.getStorageLocation());
+        }
+      }
       LOG.warn("checkDiskError got {} failed volumes - {}",
           unhealthyVolumes.size(), unhealthyVolumes);
       handleVolumeFailures(unhealthyVolumes);
+      if (!tmpDisk.isEmpty()) {
```

Comment:
some duplicate code segment in if-else statement, it would be better to optimize.

Reply:
ok